### PR TITLE
[#5066] refactor(drop-catalog): re-define drop catalog

### PR DESCRIPTION
**Pull request by mchades**
### What changes were proposed in this pull request?

 - Add an `in-use` property to the catalog with the default value of true.
 - Only empty catalog with `in-use=false` can be dropped.
 - User can use `dorce` option to drop catalog 
 - More drop catalog limitations please see the JavaDoc of `dropCatalog` in API module

### Why are the changes needed?

Fix: #5066 

### Does this PR introduce _any_ user-facing change?

yes, users now can not drop an in used catalog

### How was this patch tested?

tests added

### CODE REVIEW

File: api/src/main/java/org/apache/gravitino/Catalog.java [hunk 1]
```diff
@@ -98,6 +98,9 @@ enum CloudName {
    */
   String PROPERTY_PACKAGE = "package";
 
+  /** The property indicating the catalog is in use. */
```
Comment (yuqi1129) [inline api/src/main/java/org/apache/gravitino/Catalog.java:101]: indicating -> indicates

**Review by jerryshao — COMMENTED** [review]
Do you also need to update the doc?

File: core/src/main/java/org/apache/gravitino/catalog/CatalogManager.java [hunk 2]
```diff
@@ -474,6 +488,65 @@ public void testConnection(
...
+              Map<String, String> newProps =
+                  catalog.getProperties() == null
+                      ? new HashMap<>()
+                      : new HashMap<>(catalog.getProperties());
+              newProps.put(PROPERTY_IN_USE, "true");
+              newCatalogBuilder.withProperties(newProps);
+
+              return newCatalogBuilder.build();
+            });
+      }
```
Comment (yuqi1129) [inline core/src/main/java/org/apache/gravitino/catalog/CatalogManager.java:514]:
```suggestion
if (inUse) {
   return;
}

// Other code

```

You can use this to reduce nesting.

File: server/src/main/java/org/apache/gravitino/server/web/rest/CatalogOperations.java [hunk 3]
```diff
@@ -187,6 +187,70 @@ public Response testConnection(
...
+      @PathParam("metalake") String metalake, @PathParam("catalog") String catalogName) {
+    LOG.info("Received activate request for catalog: {}.{}", metalake, catalogName);
+    try {
+      return Utils.doAs(
+          httpRequest,
+          () -> {
+            NameIdentifier ident = NameIdentifierUtil.ofCatalog(metalake, catalogName);
+            TreeLockUtils.doWithTreeLock(
+                NameIdentifierUtil.ofMetalake(metalake),
+                LockType.READ,
```
Comment (yuqi1129) [inline server/src/main/java/org/apache/gravitino/server/web/rest/CatalogOperations.java:205]: Why do you use `READ` here? Does it mean that multiple threads can do the activated work simultaneously?

File: gradle.properties [hunk 4]
```diff
@@ -40,4 +40,4 @@ defaultScalaVersion = 2.12
 pythonVersion = 3.8
 
 # skipDockerTests is used to skip the tests that require Docker to be running.
-skipDockerTests = true
+skipDockerTests = false
```
Comment (yuqi1129) [inline gradle.properties:43]: revert it.

File: core/src/main/java/org/apache/gravitino/catalog/CatalogManager.java [hunk 5]
```diff
@@ -691,6 +807,23 @@ private CatalogWrapper createCatalogWrapper(CatalogEntity entity) {
     return wrapper;
   }
 
+  /**
+   * get the resolved properties (filter out the hidden properties and add some required default
```
Comment (yuqi1129) [inline core/src/main/java/org/apache/gravitino/catalog/CatalogManager.java:811]: Get

File: core/src/main/java/org/apache/gravitino/catalog/CatalogManager.java [hunk 6]
```diff
@@ -562,38 +627,44 @@ public Catalog alterCatalog(NameIdentifier ident, CatalogChange... changes)
...
-                EntityType.SCHEMA);
-        // If there is only one schema, it must be the default schema, because we don't allow to
-        // drop the default schema.
-        if (schemas.size() == 1) {
-          return store.delete(ident, EntityType.CATALOG, true);
-        }
+
+      if (!schemas.isEmpty()
+          && !force
+          && (!catalogEntity.getProvider().equals("kafka") || schemas.size() > 1)) {
```
Comment (yuqi1129) [inline core/src/main/java/org/apache/gravitino/catalog/CatalogManager.java:649]: Please optimize the condition, if the execution goes to `!catalogEntity.getProvider().equals("kafka") || schemas.size() > 1`,  `schemas` must be NOT empty.

File: api/src/main/java/org/apache/gravitino/exceptions/NonInUseEntityException.java [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.gravitino.exceptions;
+
+import com.google.errorprone.annotations.FormatMethod;
+import com.google.errorprone.annotations.FormatString;
+
+/** An exception thrown when operating on a non-in-use entity. */
+public class NonInUseEntityException extends GravitinoRuntimeException {
```
[the remark this answers is not in the document]
Comment (jerqi) [reply api/src/main/java/org/apache/gravitino/exceptions/NonInUseEntityException.java:25]: How about `TemporaryUnavailableException`?

File: server/src/main/java/org/apache/gravitino/server/web/Utils.java [hunk 8]
```diff
@@ -122,6 +122,28 @@ public static Response alreadyExists(String type, String message, Throwable thro
...
+        .type(MediaType.APPLICATION_JSON)
+        .build();
+  }
+
+  public static Response inUse(String message, Throwable throwable) {
+    return inUse(throwable.getClass().getSimpleName(), message, throwable);
+  }
+
+  public static Response inUse(String type, String message, Throwable throwable) {
+    return Response.status(Response.Status.CONFLICT)
```
Comment (jerqi) [inline server/src/main/java/org/apache/gravitino/server/web/Utils.java:141]: Maybe I prefer 503. Just a discussion not a suggestion.

File: clients/client-java/src/main/java/org/apache/gravitino/client/ErrorHandlers.java [hunk 9]
```diff
@@ -351,6 +359,9 @@ public void accept(ErrorResponse errorResponse) {
         case ErrorConstants.FORBIDDEN_CODE:
           throw new ForbiddenException(errorMessage);
 
+        case ErrorConstants.NON_IN_USE_ENTITY_CODE:
+          throw new NonInUseEntityException(errorMessage);
+
```
[the remark this answers is not in the document]
Comment (mchades) [reply clients/client-java/src/main/java/org/apache/gravitino/client/ErrorHandlers.java:364]: Because only drop catalog will throw in-use exception

File: core/src/main/java/org/apache/gravitino/catalog/CatalogManager.java (hunk 6 above)
Comment (mchades) [reply core/src/main/java/org/apache/gravitino/catalog/CatalogManager.java:649]: no. `schemas.size()` may equals 1

File: clients/client-python/gravitino/client/gravitino_metalake.py [hunk 10]
```diff
@@ -185,19 +185,47 @@ def alter_catalog(self, name: str, *changes: CatalogChange) -> Catalog:
...
         Args:
-            name the name of the catalog.
+            name: the name of the catalog.
+            force: whether to force drop the catalog.
 
         Returns:
-            true if the catalog is dropped successfully, false otherwise.
+            true if the catalog is dropped successfully, false if the catalog does not exist.
         """
+        params = {"force": str(force)}
```
Comment (jerryshao) [inline clients/client-python/gravitino/client/gravitino_metalake.py:198]: `str(force)` will get "True/False" or "true/false"?

File: common/src/main/java/org/apache/gravitino/dto/responses/ErrorResponse.java [hunk 11]
```diff
@@ -234,6 +234,32 @@ public static ErrorResponse alreadyExists(String type, String message, Throwable
...
 
+  /**
+   * Create a new entity in use error instance of {@link ErrorResponse}.
+   *
+   * @param type The type of the error.
+   * @param message The message of the error.
+   * @param throwable The throwable that caused the error.
+   * @return The new instance.
+   */
+  public static ErrorResponse nonInUse(String type, String message, Throwable throwable) {
```
Comment (jerryshao) [inline common/src/main/java/org/apache/gravitino/dto/responses/ErrorResponse.java:245]: "notInUse"

File: api/src/main/java/org/apache/gravitino/exceptions/EntityInUseException.java [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.gravitino.exceptions;
+
+import com.google.errorprone.annotations.FormatMethod;
+import com.google.errorprone.annotations.FormatString;
+
+/** Exception thrown when an entity is in use and cannot be deleted. */
+public class EntityInUseException extends GravitinoRuntimeException {
```
[the remark this answers is not in the document]
Comment (mchades) [reply api/src/main/java/org/apache/gravitino/exceptions/EntityInUseException.java:25]: The error message will describe the reason exactly (what entity is not in use).

If we define `CatalogInUseException`, `CatalogNotInUseException`, `MetalakeInUseException` and `MetalakeInUseException`, then we also need `SchemaNotInUseException`, `TableNotInUseException`, `TopicNotInUseException`, etc which will make the code more complicated, so I think current exception define is enough.

File: clients/client-python/gravitino/client/gravitino_metalake.py (hunk 10 above)
Comment (mchades) [reply clients/client-python/gravitino/client/gravitino_metalake.py:198]: In any case, the server side will ignore case sensitivity during query param conversion. So it's not a problem

File: server/src/main/java/org/apache/gravitino/server/web/Utils.java (hunk 8 above)
Comment (mchades) [reply server/src/main/java/org/apache/gravitino/server/web/Utils.java:141]: `SERVICE_UNAVAILABLE(503, "Service Unavailable")` means the server status, not the resource. So I think it's not suitable.

File: api/src/main/java/org/apache/gravitino/exceptions/NonInUseEntityException.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (mchades) [reply api/src/main/java/org/apache/gravitino/exceptions/NonInUseEntityException.java]: `TemporarilyUnavailableException` will introduce the conception of `Temporarily`, use `NotInUseEntityException` now

File: server/src/main/java/org/apache/gravitino/server/web/rest/CatalogOperations.java (hunk 3 above)
Comment (mchades) [reply server/src/main/java/org/apache/gravitino/server/web/rest/CatalogOperations.java]: changed to `WRITE` lock

**Comment by mchades** [comment]
> Do you also need to update the doc?

Doc will be updated in final PR

File: core/src/main/java/org/apache/gravitino/catalog/CatalogManager.java (hunk 6 above)
Comment (jerryshao) [reply core/src/main/java/org/apache/gravitino/catalog/CatalogManager.java:648]: The expression here is difficult to understand, maybe we can use multiple `if` to describe the exception message for different cases.

**Comment by mchades** [comment]
All comments are resolved, plz help to review again, thx! @yuqi1129 @jerryshao

**Comment by jerryshao** [comment]
Can you please fix the CI failure?

File: api/src/main/java/org/apache/gravitino/SupportsCatalogs.java [hunk 13]
```diff
@@ -108,12 +111,71 @@ Catalog alterCatalog(String catalogName, CatalogChange... changes)
...
+   * @param force Whether to force the drop.
+   * @return True if the catalog was dropped, false if the catalog does not exist.
+   * @throws NonEmptyEntityException If the catalog is not empty and force is false.
+   * @throws CatalogInUseException If the catalog is in use and force is false.
+   */
+  boolean dropCatalog(String catalogName, boolean force)
+      throws NonEmptyEntityException, CatalogInUseException;
+
+  /**
+   * Enable a catalog. If the catalog is already enable, this method does nothing.
```
Comment (jerryshao) [inline api/src/main/java/org/apache/gravitino/SupportsCatalogs.java:157]: "If the catalog is already enabled..."

File: core/src/main/java/org/apache/gravitino/catalog/CatalogManager.java [hunk 14]
```diff
@@ -562,38 +617,45 @@ public Catalog alterCatalog(NameIdentifier ident, CatalogChange... changes)
...
         }
       }
-      return store.delete(ident, EntityType.CATALOG);
-    } catch (NoSuchEntityException e) {
+
+      // If reached here, it implies that the catalog is not in use or force is true.
+      if (catalogInUse) {
+        // force is true, disable the catalog first
+        disableCatalog(ident);
+      }
```
Comment (jerryshao) [inline core/src/main/java/org/apache/gravitino/catalog/CatalogManager.java:649]: Since the catalog will be deleted, I don't think it is necessary to do the disable here.

File: core/src/main/java/org/apache/gravitino/listener/CatalogEventDispatcher.java [hunk 15]
```diff
@@ -169,4 +173,17 @@ public void testConnection(
     // TODO: Support event dispatching for testConnection
     dispatcher.testConnection(ident, type, provider, comment, properties);
   }
+
+  @Override
+  public void enableCatalog(NameIdentifier ident)
+      throws NoSuchCatalogException, CatalogNotInUseException {
+    // todo: support activate catalog event
```
Comment (jerryshao) [inline core/src/main/java/org/apache/gravitino/listener/CatalogEventDispatcher.java:180]: Will you do this in a different PR?

Comment (mchades) [reply core/src/main/java/org/apache/gravitino/listener/CatalogEventDispatcher.java:180]: yes

File: core/src/main/java/org/apache/gravitino/catalog/CatalogManager.java (hunk 14 above)
Comment (mchades) [reply core/src/main/java/org/apache/gravitino/catalog/CatalogManager.java:649]: The purpose of this operation is to modify the `in-use` value in the backend store

Comment (mchades) [reply core/src/main/java/org/apache/gravitino/catalog/CatalogManager.java]: removed

**Review by jerryshao — APPROVED** [review]
